refactor(ecmascript): merge constant evaluation logics

sapphi-red opened the pull request:
- merged `eval_to_*` into `eval_expression` to dedupe logics
- moved some parts of `eval_expression` to `to_number`
- changed `ConstantEvaluation` to receive `ConstantEvaluationCtx` instead of extending `IsGlobalReference` to align with other traits

Boshen commented:
### Merge activity

* **Feb 14, 7:26 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

sapphi-red committed:
refactor(ecmascript): merge constant evaluation logics (#9120)

- merged `eval_to_*` into `eval_expression` to dedupe logics
- moved some parts of `eval_expression` to `to_number`
- changed `ConstantEvaluation` to receive `ConstantEvaluationCtx` instead of extending `IsGlobalReference` to align with other traits
